_posts/2019-05-31-ucx-dgx.md

```diff
@@ -0,0 +1,531 @@
+---
+layout: widepost
+title: Experiments in High Performance Networking with UCX and DGX
+author: Matthew Rocklin
+draft: true
+tags: [python,scipy]
+theme: twitter
+---
+{% include JB/setup %}
+
+*This post is about experimental and rapidly changing software.
+Code examples in this post should not be relied upon to work in the future.*
+
+Executive Summary
+-----------------
+
+This post talks about connecting UCX, a high performance networking library, to
+Dask, a parallel Python library, to accelerate communication-heavy workloads,
+particularly when using GPUs.
+
+Additionally, we do this work on a DGX, a high-end multi-CPU multi-GPU machine
+with a complex internal network.  Working in this context was good to force
+improvements in setting up Dask in heterogeneous situations targetting
+different network cards, CPU sockets, GPUs, and so on..
+
+Moviation
+---------
+
+Many distributed computing workloads are communication-bound.
+This is common in cases like the following:
+
+1.  Dataframe joins
+2.  Machine learning algorithms
+3.  Rearranging arrays
+4.  ...
+
+Communication also increasingly becomes a bottlenck when our computation
+becomes faster, such as increasingly the case when we use GPUs for computing.
+
+Historically high performance communication was only commonly available using
+MPI, or with very custom codes.  We were trying to get close to the
+communicatino bandwidth of MPI while still maintaining the ease of
+programmability and accessiblity of a dynamic system like Dask.
+
+UCX, Python, and Dask
+---------------------
+
+To do this we wrapped up UCX with Python and then connected that to Dask.
+
+The [OpenUCX](http://www.openucx.org/) project provides a uniform API around
+various high performance networking libraries like Infiniband, traditional
+networking protocols like TCP/shared memory, and GPU-specific protocols like
+NVLink.  It is a layer beneath something like OpenMPI (the main user of OpenUCX
+today) that figures out which networking system to use.
```

Comment:
I think it would be better to specify that, despite support for TCP, shared memory, etc., the current focus is only on InfiniBand and NVLink.

Reply:
In what sense, from RAPIDS standpoint, from NVIDIA's standpoint, or from OpenUCX's standpoint? It's also worth noting issue ( https://github.com/Akshay-Venkatesh/ucx-py/issues/76 ).

Reply:
I think that we're pretty committed to TCP.  It just doesn't work well yet.  I'll add this in the future work section.